koku/api/provider/models.py

```diff
@@ -197,12 +197,187 @@ def delete(self, *args, **kwargs):
+            LOG.warning("Customer link cannot be found! Using ORM delete!")
             super().delete()
 
+    def _get_normalized_type(self):
+        """Normalize the provider type: Return lowercase type name w/o '-local'"""
+        _normalized_type = getattr(self, "_normalized_type", None)
+        if _normalized_type is None:
+            _type = self.type.lower()
+            _normalized_type = self._normalized_type = _type if "-local" not in _type else _type[: _type.index("-")]
+        return _normalized_type
+
+    def _get_sub_target_values(self, target_info, target_values):
+        sub_target = get_model(target_info["table_name"])
+        filters = {f"{target_info['column_name']}__in": target_values}
+        # This is potentially dnagerous if the sub-target is a partitioned table
+        # Care must be taken when calling this method
+        sub_target_pk = [f.name for f in sub_target._meta.fields if f.primary_key][0]
+        values = [r[sub_target_pk] for r in sub_target.objects.filter(**filters).values(sub_target_pk)]
+        return values
+
+    def _cascade_delete(self, target_table=None, target_values=None, public_schema="public", seen=None):
+        if seen is None:
+            seen = {"public", "api_providerinfrastructuremap", self.pk}
+        if target_table is None:
+            target_table = self._meta.db_table
+        if target_values is None:
+            target_values = [self.pk]
+        _normalized_type = self._get_normalized_type()
+        # If one of the targets is in this tuple, then we must recursively call _cascade_delete
+        # because this is a second level of indirection to get to data:
+        # Provider ---> <branch-table> ---> x...
+        # Make sure that any table(s) in the public schema are first.
+        _cascade_branch_tables = (
+            "reporting_common_costusagereportmanifest",
+            f"reporting_{_normalized_type}costentrybill",
+            f"reporting_{_normalized_type}meter",
+            f"reporting_{_normalized_type}usagereportperiod",
+        )
+
+        for target_info in self._get_linked_table_names(_normalized_type, target_table, public_schema):
+            _target_vals = tuple(target_info[k] for k in sorted(target_info))
+            if _target_vals in seen:
+                continue
+
+            seen.add(_target_vals)
+            if target_info["table_name"] == "api_providerinfrastructuremap":
+                self._set_infrastructure_id_null()
+
+            if target_info["table_name"] in _cascade_branch_tables:
+                # Keep the slice up-tp-date with the number of public schema table names in _cascade_branch_tables
+                public_schema = (
+                    self.customer.schema_name if target_info["table_name"] in _cascade_branch_tables[1:] else "public"
+                )
+                LOG.info(f"DELETE CASCADE BRANCH TO {target_info['table_name']}")
```

Comment:
Let's change these to DEBUG statements. Otherwise there is a _lot_ of stuff logged. I think all log statements in `_cascade_delete` should be debug, and the ones in `delete` should be info.